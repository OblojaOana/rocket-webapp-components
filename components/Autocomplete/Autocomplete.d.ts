
export interface OptionTypeBase {
  [key: string]: any;
}

export type OptionsType<OptionType> = ReadonlyArray<OptionType>;

export interface AutocompleteProps {
  /**
   * The array of options from which the client can select a value.
   * @default []
   */
  options?: []
  /**
  * Function that returns a promise, which is the set of options to be used once the promise resolves.
  */
  loadOptions(inputValue: string, callback: (options: OptionsType<OptionType>) => void): void;
  /**
    * The selected value from list of options.
    * @default null
    */
  value?: Object | [] | number | string | boolean
  /**
  * Handle change events on the select.
  * This property is required.
  */
  onChange: () => void
  /**
  * Handle the menu opening.
  */
  onMenuOpen?: () => void
  /**
  * If true, the user can select multiple values from list.
  * @default false  
  */
  isMultiSelection?: boolean
  /**
  * If true, the user can clear the selected value.
  * @default false  
  */
  isClearable?: boolean
  /**
  * If true, the search functionality is enabled.
  * @default false  
  */
  isSearchable?: boolean
  /**
  * If true, the Select is disabled.
  * @default false  
  */
  disabled?: boolean
  /**
  *@todo
  * @default false
  */
  simpleValue?: boolean
  /**
  * Label to be displayed in the heading component.
  */
  label?: string
  /**
  * The key of values from options.
  * @default "id"
  * @TODO
  */
  valueKey?: string
  /**
  * @TODO
  * @default "name"
  */
  labelKey?: string
  /**
  * If true, the helper text is displayed when an error pops up.
  * @default false
  */
  error?: boolean
  /**
  * The value of label when a new option is added.
  */
  createdLabel?: string
  /**
  * The default set of options to show before the user starts searching. When set to true, the results for loadOptions('') will be autoloaded.
  */
  defaultOptions?: boolean | Object[]
}
/**
 *
 * Demos:
 *
 * - https://bit.dev/totalsoft_oss/react-mui/autocomplete
 *
 * 
 */
export default function Autocomplete(props: AutocompleteProps): JSX.Element;