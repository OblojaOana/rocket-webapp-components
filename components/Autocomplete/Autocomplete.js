/* eslint-disable react/prop-types */
import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import AsyncCreatableSelect from 'react-select/async-creatable';
import AsyncSelect from 'react-select/async';
import { Paper, MenuItem, TextField, ListItem, ListItemIcon, ListItemText, Checkbox, makeStyles, InputAdornment } from '@material-ui/core';
import Typography from '../Typography';
import autoCompleteStyles from "./autocompleteStyle"
import { isArray } from 'util';
import Search from '@material-ui/icons/Search';
import { curry, prop } from 'ramda';
import i18next from 'i18next';

const useStyles = makeStyles(autoCompleteStyles);

function NoOptionsMessage({ selectProps, innerProps, children }) {
  return (
    <Typography
      color="textSecondary"
      className={selectProps.classes.noOptionsMessage}
      {...innerProps}
    >
      {children}
    </Typography>
  );
}

function inputComponent({ inputRef, ...props }) {
  return <div ref={inputRef} {...props} />;
}

function controlHasValue(selectProps) {
  if (selectProps.inputValue.length > 0) {
    return true;
  }
  if (selectProps.value === null || selectProps.value === undefined) {
    return false;
  }
  if (isArray(selectProps.value) && selectProps.value.length === 0) {
    return false;
  }

  return true;
}

function Control({ selectProps, innerRef, children, innerProps }) {

  const inputProps = {
    inputComponent,
    inputProps: {
      className: selectProps.classes.input,
      inputRef: innerRef,
      children: children,
      ...innerProps,
    },
    endAdornment: (
      <InputAdornment>
        <Search />
      </InputAdornment>)
  }

  if (!selectProps.isSearchable) {
    delete inputProps.endAdornment;
  }

  return (
    <TextField
      fullWidth
      InputProps={inputProps}
      InputLabelProps={{
        shrink: controlHasValue(selectProps),
        style: { fontSize: 'inherit' }
      }}
      value={selectProps.value ? selectProps.getOptionValue(selectProps.value) : null}
      {...selectProps.textFieldProps}
    />
  );
}

function Option({ innerRef, isFocused, isSelected, innerProps, children }) {
  return (
    <MenuItem
      buttonRef={innerRef}
      selected={isFocused}
      component="div"
      style={{
        fontWeight: isSelected ? 500 : 400,
        fontSize: "14px"
      }}
      {...innerProps}
    >
      {typeof children === 'function' ? children() : children}
    </MenuItem>
  );
}

const MultiOption = props => (
  <ListItem dense button selected={props.isSelected} {...props.innerProps}>
    <ListItemIcon>
      <Checkbox
        edge="start"
        checked={props.isSelected}
        tabIndex={-1}
        disableRipple
      />
    </ListItemIcon>
    <ListItemText primary={props.label} />
  </ListItem>

);

function Placeholder({ selectProps, innerProps, children }) {
  return (
    <Typography
      color="textSecondary"
      className={selectProps.classes.placeholder}
      {...innerProps}
    >
      {children}
    </Typography>
  );
}

function SingleValue({ selectProps, innerProps, children }) {
  return (
    <Typography className={selectProps.classes.singleValue} {...innerProps}>
      {typeof children === 'function' ? children() : children}
    </Typography>
  );
}

function ValueContainer({ selectProps, children }) {
  return <div className={selectProps.classes.valueContainer}>{children}</div>;
}

function MultiValueDisplay(text) {
  return (
    <span key={text}>
      {text}
    </span>
  );
}

function MultiValueContainer({ selectProps, children }) {

  let elements = Object.assign([], children);
  if (children.length > 0 && children[0] && children[0].length > 1) {
    elements[0] = MultiValueDisplay(`(${children[0].length} selected)`);
  }

  return <div className={selectProps.classes.valueContainer}>{elements}</div>;
}


function MultiValue(props) {
  return (
    <span>{props.children}</span>
  );
}

function Menu({ selectProps, innerProps, children }) {
  return (
    <Paper square className={selectProps.classes.paper} {...innerProps}>
      {children}
    </Paper>
  );
}

function IndicatorSeparator() {
  return (<span />);
}

function getSimpleValue(options, value, valueKey, isMultiSelection) {
  const option = options.filter(a => a[valueKey] === value)[0];

  if (isMultiSelection) {
    return options.filter(a => value && value.includes(a[valueKey])) || [];
  }
  return option || null;
}

function DropdownIndicator() {
  return (<span />);
}

const formatCreateLabel = curry((createdLabel, text) => i18next.t(createdLabel, { text }))

function Autocomplete({ options, defaultOptions, loadOptions, onChange,
  creatable, onMenuOpen, value, isMultiSelection, isClearable,
  isSearchable, disabled, simpleValue, label, valueKey, labelKey, error, helperText, createdLabel, ...other }) {
  const classes = useStyles();

  const handleOnChange = useCallback(inputValue => {
    if (simpleValue && inputValue) {
      if (isMultiSelection) {
        return onChange(inputValue.map(a => a[valueKey]))
      } else {
        return onChange(inputValue[valueKey])
      }
    }

    return onChange(inputValue)
  }, [isMultiSelection, onChange, simpleValue, valueKey])

  const getNewOptionData = useCallback((inputValue, optionLabel) => ({
    [labelKey]: optionLabel,
    [valueKey]: undefined
  }), [labelKey, valueKey])

  const Comp = loadOptions ? creatable ? AsyncCreatableSelect : AsyncSelect : Select;
  const loadOptionsAsync = useCallback(input => {
    return loadOptions(input)
  }, [loadOptions])

  const components = {
    Control,
    Menu,
    MultiValue,
    NoOptionsMessage,
    Option: isMultiSelection ? MultiOption : Option,
    Placeholder,
    SingleValue,
    ValueContainer: isMultiSelection ? MultiValueContainer : ValueContainer,
    IndicatorSeparator,
  }

  if (isSearchable) {
    components.DropdownIndicator = DropdownIndicator;
  }

  return (
    <Comp
      {...other}
      classes={classes}
      styles={classes.selectStyles}
      options={options}
      loadOptions={loadOptionsAsync}
      defaultOptions={defaultOptions}
      components={components}
      closeMenuOnSelect={isMultiSelection ? false : true}
      value={simpleValue ? getSimpleValue(options, value, valueKey, isMultiSelection) : value}
      onChange={handleOnChange}
      onMenuOpen={onMenuOpen}
      textFieldProps={{
        label: label,
        error: error,
        helperText: helperText
      }}
      multiValueRemove={false}
      creatable={creatable}
      hideSelectedOptions={false}
      placeholder={null}
      isMulti={isMultiSelection}
      isClearable={isClearable}
      isSearchable={isSearchable}
      isDisabled={disabled}
      getOptionLabel={prop(labelKey)}
      getOptionValue={prop(valueKey)}
      formatCreateLabel={formatCreateLabel(createdLabel)}
      getNewOptionData={getNewOptionData}
    />
  )
}

Autocomplete.defaultProps = {
  options: [],
  isMultiSelection: false,
  isClearable: false,
  isSearchable: false,
  disabled: false,
  simpleValue: false,
  valueKey: "id",
  labelKey: "name",
  label: " ",
  error: false,
  value: null,
  creatable: false
};

Autocomplete.propTypes = {
  /**
  * The array of options from which the client can select a value.
  * @default []
  */
  options: PropTypes.array,
  /**
  * Function that returns a promise, which is the set of options to be used once the promise resolves.
  */
  loadOptions: PropTypes.func,
  /**
  * The selected value from list of options.
  * @default null
  */
  value: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.number,
    PropTypes.string,
    PropTypes.bool
  ]),
  /**
  * Handle change events on the select.
  */
  onChange: PropTypes.func.isRequired,
  /**
  * Handle the menu opening.
  */
  onMenuOpen: PropTypes.func,
  /**
  * If true, the user can select multiple values from list.
  * @default false  
  */
  isMultiSelection: PropTypes.bool,
  /**
  * If true, the user can clear the selected value.
  * @default false  
  */
  isClearable: PropTypes.bool,
  /**
  * If true, the search functionality is enabled.
  * @default false  
  */
  isSearchable: PropTypes.bool,
  /**
  * If true, the Select is disabled.
  * @default false  
  */
  disabled: PropTypes.bool,
  /**
  * @default false
  * @TODO
  */
  simpleValue: PropTypes.bool,
  /**
  * Label to be displayed in the heading component.
  */
  label: PropTypes.string,
  /**
  * The key of values from options.
  * @default "id"
  * @TODO
  */
  valueKey: PropTypes.string,
  /**
  * @default "name"
  * @TODO
  */
  labelKey: PropTypes.string,
  /**
  * If true, the helper text is displayed when an error pops up.
  * @default false
  */
  error: PropTypes.bool,
  /**
  * The value of label when a new option is added.
  */
  createdLabel: PropTypes.string,
  /**
  * The default set of options to show before the user starts searching. When set to true, the results for loadOptions('') will be autoloaded.
  */
  defaultOptions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array])
};

export default Autocomplete;