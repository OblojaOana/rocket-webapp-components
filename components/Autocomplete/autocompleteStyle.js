import { emphasize } from '@material-ui/core/styles/colorManipulator';
import styles from '../../assets/jss/styles';
import { theme as defaultTheme } from '../../themes/defaultTheme';

export const selectedColor = (theme, inputSelectedColor) => {
  return { color: inputSelectedColor ? inputSelectedColor : theme.autoComplete.selectedInputColor }
}

const autocompleteStyle = theme => {
  const {
    dangerColor,
    successColor,
    defaultFont
  } = styles(theme);


  return ({
    root: {
      flexGrow: 1,
      height: 250,
    },
    input: {
      ...defaultFont,
      display: 'flex',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis'
    },
    label: {
      ...defaultFont
    },
    labelError: {
      color: dangerColor[0] + " !important"
    },
    labelSuccess: {
      color: successColor[0] + " !important"
    },
    valueContainer: {
      ...theme.autoComplete?.valueContainer || defaultTheme.autoComplete.valueContainer,
      '& input': {
        ...theme.autoComplete?.input || defaultTheme.autoComplete.input
      }
    },
    chip: {
      margin: `${theme.spacing(0.5)}px ${theme.spacing(0.25)}px`,
    },
    chipFocused: {
      backgroundColor: emphasize(
        theme.palette.type === 'light' ? theme.palette.grey[300] : theme.palette.grey[700],
        0.08,
      ),
    },
    noOptionsMessage: {
      padding: `${theme.spacing()}px ${theme.spacing(2)}px`,
    },
    singleValue: {
      ...defaultFont,
      color: theme.palette?.text?.primary || defaultTheme.palette.text.primary
    },
    placeholder: {
      ...defaultFont,
      position: 'absolute',
      left: 2
    },
    paper: {
      position: 'absolute',
      zIndex: 1,
      marginTop: theme.spacing(),
      left: 0,
      right: 0,
    },
    divider: {
      height: theme.spacing(2),
    },
    selectStyles: {
      singleValue: base => ({
        ...base,
        color: theme.palette?.text?.primary || defaultTheme.palette.text.primary
      }),
      input: base => ({
        ...base,
        color: theme.palette?.text?.primary || defaultTheme.palette.text.primary
      })
    }
  })
};

export default autocompleteStyle;