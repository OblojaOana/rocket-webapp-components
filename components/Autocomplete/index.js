import Autocomplete from "../../core/Autocomplete";
export default Autocomplete;