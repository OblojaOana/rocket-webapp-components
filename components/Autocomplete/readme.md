##### Copy this snapshot in the overview panel
 ```js
import React, { useState } from 'react';
import Autocomplete from '@bit/totalsoft_oss.react-mui.autocomplete';
import curry from 'lodash.curry';

const arr = [{ id: '1', name: 'Approved' }, { id: '2', name: 'Cancelled' }]

const App = () => {
	const [localFilters, setLocalFilters] = useState(' ');
	const handleFilterPropertyChange = curry((prop, value) =>
		setLocalFilters(prevFilters => ({ ...prevFilters, [prop]: value }))
	);

	return (
		<Autocomplete
			label={"Status"}
			options={arr}
			fullWidth
			simpleValue={true}
			isClearable={true}
			value={localFilters.statusId || ''}
			onChange={handleFilterPropertyChange('id')}
		/>
	)
}

export default (
	<App />
)
 ```