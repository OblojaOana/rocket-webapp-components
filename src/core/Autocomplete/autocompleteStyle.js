import { emphasize } from '@material-ui/core/styles/colorManipulator';
import styles from '@bit/totalsoft_oss.react-mui.styles';
import { theme as DefaultTheme } from '@bit/totalsoft_oss.react-mui.themes.default-theme';

const autocompleteStyle = theme => {
  const { dangerColor, successColor, defaultFont } = styles(theme);

  return ({
    root: {
      flexGrow: 1,
      height: 250,
    },
    input: {
      ...defaultFont,
      display: 'flex',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis'
    },
    label: {
      ...defaultFont
    },
    labelError: {
      color: dangerColor[0] + " !important"
    },
    labelSuccess: {
      color: successColor[0] + " !important"
    },
    valueContainer: {
      ...theme.autoComplete?.valueContainer || DefaultTheme.autoComplete.valueContainer,
      '& input': {
        ...theme.autoComplete?.input || DefaultTheme.autoComplete.input
      }
    },
    chip: {
      margin: `${theme.spacing(0.5)}px ${theme.spacing(0.25)}px`,
    },
    chipFocused: {
      backgroundColor: emphasize(
        theme.palette.type === 'light' ? theme.palette.grey[300] : theme.palette.grey[700],
        0.08,
      ),
    },
    noOptionsMessage: {
      padding: `${theme.spacing()}px ${theme.spacing(2)}px`,
    },
    singleValue: {
      ...defaultFont
    },
    placeholder: {
      ...defaultFont,
      position: 'absolute',
      left: 2
    },
    paper: {
      position: 'absolute',
      zIndex: 1,
      marginTop: theme.spacing(),
      left: 0,
      right: 0,
    },
    divider: {
      height: theme.spacing(2),
    },
    selectStyles: {
      input: base => ({
        ...base,
        color: theme.palette?.text?.primary || DefaultTheme.palette.text.primary
      }),
    }
  })
};

export default autocompleteStyle;