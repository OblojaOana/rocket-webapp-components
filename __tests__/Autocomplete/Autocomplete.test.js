import React from 'react';
import renderer from 'react-test-renderer';
import Autocomplete from '../../components/Autocomplete/Autocomplete';

const arr = [{ id: '1', name: 'Approved' }, { id: '2', name: 'Cancelled' }]

const handleFilterPropertyChange = () => (_, _) => { }

describe('Autocomplete', () => {
    it('snapshot', () => {
        const component = renderer.create(<Autocomplete
            label={"Status"}
            options={arr}
            fullWidth
            simpleValue
            isClearable
            value={''}
            onChange={handleFilterPropertyChange('id')}
        />);

        const json = component.toJSON();
        expect(json).toMatchSnapshot();
    });
});