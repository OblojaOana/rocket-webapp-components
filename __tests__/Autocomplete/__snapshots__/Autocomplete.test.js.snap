// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete snapshot 1`] = `
<div
  className=" css-2b097c-container"
  onKeyDown={[Function]}
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
      style={
        Object {
          "fontSize": "inherit",
        }
      }
    >
      Status
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <div
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input makeStyles-input-2"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        required={false}
        type="text"
        value={null}
      >
        <div
          className="makeStyles-valueContainer-6"
        >
          <span
            className="MuiTypography-root makeStyles-placeholder-11 MuiTypography-colorTextSecondary"
          />
          <input
            aria-autocomplete="list"
            className="css-62g3xt-dummyInput"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            tabIndex="0"
            value=""
          />
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span />
          <div
            aria-hidden="true"
            className=" css-tlfecz-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <svg
              aria-hidden="true"
              className="css-6q0nyr-Svg"
              focusable="false"
              height={20}
              viewBox="0 0 20 20"
              width={20}
            >
              <path
                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
